seisbench/models/deepdenoiser.py

```diff
@@ -284,3 +283,454 @@ def forward(self, x, mid):
+            out_channels = 2**i * self.filters_root
+            self.up_blocks.append(
+                TransposeConvBlock(
+                    in_channels=in_channels,
+                    out_channels=out_channels,
+                    kernel_size=self.kernel_size,
+                    stride=self.stride,
+                    drop_rate=self.drop_rate,
+                    use_bias=self.use_bias,
+                )
+            )
+            if self.attention:
+                self.attention_gates.append(
+                    AttentionGate(
+                        in_channels_encoder=out_channels,
+                        in_channels_decoder=out_channels,
+                        inter_channels=out_channels // 2,
+                    )
+                )
+            else:
+                self.attention_gates.append(None)
+
+            self.decoder_blocks.append(
+                ConvBlock(
+                    in_channels=out_channels,
+                    out_channels=out_channels,
+                    kernel_size=self.kernel_size,
+                    drop_rate=self.drop_rate,
+                    use_bias=self.use_bias,
+                )
+            )
+
+        self.output_conv = nn.Conv2d(
+            in_channels=self.filters_root, out_channels=self.in_channels, kernel_size=1
+        )
+
+    def forward(self, x):
+        enc_features = []  # List to store encoder feature for skip connections
+
+        # Encoder
+        for i in range(self.depth):
+            x = self.encoder_blocks[i](x)
+            enc_features.append(x)
+            if i < self.depth - 1:
+                x = self.down_blocks[i](x)
+
+        # Decoder
+        for i in range(self.depth - 2, -1, -1):
+            x = self.up_blocks[self.depth - 2 - i](x)
+
+            # Pad if needed to match encoder feature size
+            diff_y = enc_features[i].size(2) - x.size(2)
+            diff_x = enc_features[i].size(3) - x.size(3)
+            x = F.pad(
+                x,
+                [diff_x // 2, diff_x - diff_x // 2, diff_y // 2, diff_y - diff_y // 2],
+            )
+
+            if self.attention:  # Apply attention gate to encoder feature
+                skip = self.attention_gates[self.depth - 2 - i](enc_features[i], x)
+                x = x + skip  # gated skip connection
+            else:  # No attention gate
+                x = x + enc_features[i]  # Skip connection
+            x = self.decoder_blocks[self.depth - 2 - i](x)
+
+        x = self.output_conv(x)
+
+        # Apply out activation function
+        if self.output_activation:
+            x = self.output_activation(x)
+
+        return x
+
+    def annotate_batch_pre(
+        self, batch: torch.Tensor, argdict: dict[str, Any]
+    ) -> torch.Tensor:
+        """
+        Does preprocessing for prediction
+        """
+        # STFT of each batch and component
+        self.norm_factors = np.empty(shape=batch.shape[0])
```

Comment:
I think it would be better to avoid writing to a class variable here. Just add `norm_factors` to the piggyback by returning `noisy_input, (noisy_stft, norm_factors)` at the bottom.